EIPS/eip-663.md

```diff
@@ -58,11 +62,21 @@ Since this instruction depends on an immediate argument encoding, it can only be
 
 ### Size of immediate argument
 
-A 16-bit size was considered to accommodate the full stack space of 1024 items, however:
+For `DUPN` a 16-bit size was considered to accommodate the full stack space of 1024 items, however:
 
 1. that would require an additional restriction/check (`n < 1024`)
 2. the 256 depth is a large improvement over the current 16 and the overhead of an extra byte would make it less useful
 
+Similarly for `EXCHANGE`, the proposed scheme allows addressing of 32 items.
+
+### Gas cost
+
+The gas cost for these operations is the same as for existing `DUP*` and `SWAP*` instructions, because they are just implemented as pointer swaps.
+
+### Exchange vs extending swap
+
+As mentioned before, `EXCHANGE` is important to compilers implementing stack scheduling algorithms. Specifically, in the case that a stack item is scheduled to be consumed deeper in the stack (for instance, the 3rd item in the stack needs to be moved into 2nd position in order to be consumed by the next operation), that currently takes three instructions, `SWAP2 SWAP3 SWAP2`. However, in the EVM implementation, the implementation is just a pointer swap, so it could be implemented in a single instruction at no extra runtime cost to the client.
+
```

Comment:
This argument here also technically doesn't hold up to scrutiny.
Proper EVM stack scheduling can avoid situations in which you need to exchange two slots neither of which are at the stack top. I.e. if you find yourself in the situation of having to swap the 3rd and 2nd position on stack, then you allocated an inefficient position for the values on stack when you produced them in the first place. I've given a talk at Devconnect Amsterdam about the topic, if you're interested.
Conversely, all values in the EVM are *produced* at the stack top and it thereby cannot be avoided to swap the stack top with a deeper position to create optimal stack layouts in the first place.

So, technically, `SWAPN` is actually the more efficient choice, since the more common and unavoidable operation on stack top requires only two bytes instead of three even for deep access, while conversely patterns like `SWAP2 SWAP3 SWAP2` can generally be avoided by proper stack scheduling at the compiler level.

That being said, having written stack scheduling like that myself, I know that it is not the most trivial thing to do - that's why I've not immediately tried to shut down the idea of `EXCHANGE`: it does allow for simpler more naive stack scheduling albeit at the cost of an increase in code size cost globally compared to optimal scheduling.

But either way, `DUPN` and `EXCHANGE` are not a replacement of `SWAPN`, so it would have to be `EXCHANGE2` as well, no matter what.

Reply:
> generally be avoided by proper stack scheduling at the compiler level.

the situation can definitely be improved with better algorithms, but optimal stack scheduling, especially across basic blocks and subroutines, is hard. (although i don't have a proof, i suspect it is NP-complete, in the same way that register allocation is). and, even if such a theoretically optimal stack scheduler were available, it is sometimes simply not possible in the general case to schedule things to be in the right order ahead of time, for instance, when side effects are interleaved with stack item production.

> But either way, DUPN and EXCHANGE are not a replacement of SWAPN, so it would have to be EXCHANGE2 as well, no matter what.

having things too deep in the stack is also solvable by better memory spilling algorithms, so i don't think phrasing this choice in absolute terms is very constructive here. but sure, if `EXCHANGE` + `EXCHANGE2` are not considered as taking up too many slots, why not.